kyuubi-server/src/main/scala/org/apache/kyuubi/engine/EngineRef.scala

```diff
@@ -125,12 +128,34 @@ private[kyuubi] class EngineRef private(conf: KyuubiConf, user: String, sessionI
       }
   }
 
-  private def create(zkClient: CuratorFramework): (String, Int) = tryWithLock(zkClient) {
-    // TODO: improve this after support engine pool. (KYUUBI #918)
-    var engineRef = getServerHost(zkClient, engineSpace)
-    // Get the engine address ahead if another process has succeeded
-    if (engineRef.nonEmpty) return engineRef.get
+  def getOrCreate(zkClient: CuratorFramework): (String, Int) = tryWithLock(zkClient) {
+
+    // USER share level support engine pool
+    if (shareLevel.equals(USER)) {
+
+      // create engine space
+      createEngineSpaceIfNotExists(zkClient, engineSpace, conf.get(HA_ZK_ENGINE_POOL_SIZE))
+
+      // create engine if not reach pool size
+      val notFull = checkEnginePoolCapacity(zkClient, engineSpace)
+      if (notFull) {
+        createInternal(zkClient)
+      } else {
+        val engineRef = getEngineByPolicy(zkClient, engineSpace, providePolicy)
+        if (engineRef.nonEmpty) return engineRef.get
+        else throw new Exception("Unexpected exception!!!")
```

Comment:
Avoid using raw `Exception`, pls pick a specific `KyuubiException` or create a new subclass if there is no suitable one.